src/ripple/app/tx/impl/Payment.cpp

```diff
@@ -160,6 +161,53 @@ class Payment
                 "No ripple direct specified for XRP to XRP.";
             return temBAD_SEND_XRP_NO_DIRECT;
         }
+        if (bDeliverMin)
+        {
+            auto const deliverMin = mTxn.getFieldAmount (sfDeliverMin);
+            if (!isLegalNet (deliverMin) || deliverMin <= zero)
+            {
+                if (m_journal.trace) m_journal.trace <<
+                    "Malformed transaction: bad deliver min amount: " <<
+                    deliverMin.getFullText ();
+
+               return temBAD_AMOUNT;
+            }
+            if (deliverMin.getCurrency () != uDstCurrency)
+            {
+                if (m_journal.trace) m_journal.trace <<
+                    "Malformed transaction: deliver min currency"
+                    "differs from dst: " <<
+                        maxSourceAmount.getFullText ();
+
+                return temBAD_AMOUNT;
+            }
+            if (deliverMin.getIssuer () != saDstAmount.getIssuer ())
+            {
+                if (m_journal.trace) m_journal.trace <<
+                    "Malformed transaction: deliver min issuer"
+                    "differs from dst: " <<
+                        maxSourceAmount.getFullText ();
+
+                return temBAD_AMOUNT;
+            }
+            if (deliverMin > saDstAmount)
+            {
+                if (m_journal.trace) m_journal.trace <<
+                    "Malformed transaction: deliver min greater than dst: " <<
+                        maxSourceAmount.getFullText();
+
+                return temBAD_AMOUNT;
+            }
+            if (bMax && deliverMin > maxSourceAmount)
```

Comment:
This doesn't look right. These may be different currencies, right? Does it make sense to check `maxSourceAmount`?